x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchTemplate.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.search;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public class SearchTemplate implements ToXContentObject, Writeable {
+    private final Script script;
+
+    public SearchTemplate(StreamInput in) throws IOException {
+        this.script = in.readOptionalWriteable(Script::new);
+    }
+
+    public SearchTemplate(Script script) {
+        this.script = script;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(SearchApplication.TEMPLATE_SCRIPT_FIELD.getPreferredName(), script);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalWriteable(script);
+    }
+
+    public static SearchTemplate parse(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    private static final ConstructingObjectParser<SearchTemplate, Void> PARSER = new ConstructingObjectParser<>(
+        "search_template",
+        p -> new SearchTemplate((Script) p[0])
+    );
+
+    static {
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> Script.parse(p), SearchApplication.TEMPLATE_SCRIPT_FIELD);
```

Comment:
The default language is `painless` when using `Script.parse(p)`. We can use `Script.parse(p, Script.DEFAULT_TEMPLATE_LANG)` to default to `mustache`. We should also validate after parsing that the language (`lang`) is not set to something else?